numpy/lib/index_tricks.py

```diff
@@ -589,6 +590,414 @@ def __iter__(self):
+    be acheived readily through:
+
+    >>> for i in np.ndrange((3, 2)[::-1]):
+    ...     i = i[::-1]
+    ...     print(i)
+    (0, 0)
+    (1, 0)
+    (2, 0)
+    (0, 1)
+    (1, 1)
+    (2, 1)
+
+    or
+
+    >>> for i_f in (i_c[::-1] for i_c in np.ndrange((3, 2)[::-1])):
+    ...     print(i_f)
+    (0, 0)
+    (1, 0)
+    (2, 0)
+    (0, 1)
+    (1, 1)
+    (2, 1)
+
+    The conceptual ``r`` can be obtained by making a call to ``np.asarray``.
+    We strongly against this method of array creation since it is slow.
+
+    >>> np.asarray(np.ndrange((4, 4))[::2, 1:-1])
+    array([[(0, 1), (0, 2)],
+           [(2, 1), (2, 2)]], dtype=object)
+
+    """
+
+    __slots__ = ['_start', '_stop', '_step', '_ranges']
+
+    def __init__(self, start=None, stop=None, step=None):
+        if stop is None:
+            stop = start
+            start = None
+
+        if stop is None:
+            raise ValueError('``stop`` must be provided.')
+
+        # Is there a function that deals with this input sanitization for me?
+        if not isinstance(stop, _nx.collections_abc.Iterable):
+            stop = (stop,)
+        elif not isinstance(stop, tuple):
+            stop = tuple(stop)
+
+        if start is None:
+            start = (0,) * len(stop)
+        elif not isinstance(start, _nx.collections_abc.Iterable):
+            start = (start,)
+        elif not isinstance(start, tuple):
+            start = tuple(start)
+
+        if step is None:
+            step = (1,) * len(stop)
+        elif not isinstance(step, _nx.collections_abc.Iterable):
+            step = (step,)
+        elif not isinstance(step, tuple):
+            step = tuple(step)
+
+        if len(start) != len(stop):
+            raise ValueError(
+                "``start`` and ``stop`` must have the same length.")
+
+        if len(step) != len(stop):
+            raise ValueError(
+                "``start``, ``stop``, and ``step`` must have the same length.")
+
+        self._start = start
+        self._stop = stop
+        self._step = step
+        # The problem with using xrange objects in python2 is that
+        # they don't support being indexed by slices.
+
+        # Create the range objects here so that in case of invalid parameters
+        # ndrange fails early
+        self._ranges = tuple(range(*r) for r in zip(start, stop, step))
+
+    @property
+    def start(self):
+        """
+        The value of the start parameter (or ``(0,) * len(stop)`` if the
+        parameter was not supplied)
+        """
+        return self._start
+
+    @property
+    def stop(self):
+        """
+        The provided ``stop`` parameter as a tuple.
+        """
+        return self._stop
+
+    @property
+    def step(self):
+        """
+        Tuple containing the ``step`` parameter along each axis.
+        """
+        return self._step
+
+    def count(self, x):
+        """
+        Return number of occurrences of the provided tuple.
```

Comment:
If we're going to add `count` to ndrange, then I think we should add it to `ndarray` first.

Reply:
This is to match the collections abc API. Not really the ndarray API